### ISSUE TITLE
CI: move Azure jobs to GitHub Actions

### ISSUE THREAD

**Issue by tupui** [scipy/scipy#15814]
As discussed during the last community meeting, we want to start moving away from Azure.

This would allow to have faster CI (Azure is slow), but also improve our workflow (less clicking) and potentially reduce our consumption–as for new contributors, GH requires a maintainer to approve before running any CI.

Apart from putting all our eggs in the same basket issue, there is also a concern about skipping some CI. As we move everything to 1 or 2 platforms, we can take this opportunity to rethink the skipping strategy. I propose to separate things in logical modules instead of platforms. Ex we could have a `minimal` group that would do things like linter, ref guide, doc, one full job. We could even consider opt-in groups.

**Comment by rgommers** [comment]
Let me just point out this work, which looks in detail at redesigning the CI matrix for SciPy: https://labs.quansight.org/blog/2021/10/re-engineering-cicd-pipelines-for-scipy/. That is all still valid; the exact matrix contents can be evolved, but let's not start from scratch.

For now, moving 1-2 jobs that are timing out on Azure to GitHub Actions and seeing if Windows jobs are faster there sounds great though.

**Comment by AnirudhDagar** [comment]
As discussed offline, the other job that is to be moved from Azure over to GH Actions is `Linux_Python_38_32bit_full`. I'll send a PR for that soon.

**Comment by asmijafar20** [comment]
Hey. I would like to work on this issue. I was experienced with CI/CD pipelines as I implemented the CI/CD pipelines throughout the Moja Global community during my Outreachy internship. If this issue is still up, I would like to take it and want to know more about it and discuss it further. Thanks!

**Comment by AnirudhDagar** [comment]
Hi @asmijafar20, @tupui already finished one of the CI job moves. I've finished the other one. Just need to fix a small thing and raise the PR. Feel free to pick-up other CI issues which are also interesting. You might be interested in the follow-ups that will be made after #15959. Overall #15489 might also be something that interests you. :))

**Comment by tupui** [comment]
@asmijafar20 
Before starting anything,I would still suggest you detail your plan here. We don't have this public yet, but there is a state we would want to move towards.

**Comment by asmijafar20** [comment]
> @asmijafar20 Before starting anything,I would still suggest you detail your plan here. We don't have this public yet, but there is a state we would want to move towards.

Sure.

**Comment by rgommers** [comment]
Okay, it seems like we're ready to do this! I will look at sending a PR today to reduce the current load on our GitHub Actions usage by removing some things (`setup.py` and other) that we no longer need. That makes enough room for new jobs so that we can get rid of Azure completely. I think these are:

- [ ] a lint job (separate, fast job). there are 3 checks: `tools/lint.py`, `tools/unicode-check.py`, `tools/check_test_name.py`
- [ ] a job with our minimum supported GCC version (8.x right now) and also using `export PYTHONOPTIMIZE=2`
- [ ] a job with pre-releases or the `main` branch of all our dependencies. Basically that is using `"--pre --upgrade --timeout=60 -i $PRE_WHEELS` where the env var points at https://pypi.anaconda.org/scipy-wheels-nightly/simple
- [ ] regular Windows jobs. I think we need four in total (extend `.github/workflows/windows.yml`, where we already have one): 
    - [ ] 32-bit Python 3.9, using `dev.py`, Python 3.9, `'full'` test run, minimum supported `numpy` version
    - [ ] 64-bit Python 3.9, using `pip install .`, Python 3.9, default (`'fast'`) test run
    - [ ] 32-bit Python 3.10, using `dev.py`, `'fast'` test run
    - [ ] 64-bit Python 3.11, using `python -m build && pip install dist/scipy*.whl` `'full'` test run (add ILP64 to this later on)


The jobs I think we can drop (checked off if there's nothing left to do):

- [ ] the `asv` benchmarking job (see `asv_check` in `ci/azure-travis-template.yaml`)~ no need, already running on Circle CI
- [ ] Codecov - let's keep it on our radar, but I propose not migrating this but just dropping it for now. They had yet another security incident yesterday ([link](https://community.codecov.com/t/codecov-yanked-from-pypi-all-versions/4259), didn't affect us luckily), and are generally plain untrustworthy.
- [ ] a separate "use sdist" job. This can easily be weaved into an existing job, it's a one-liner.

Leave for last:

- [ ] the 32-bit Linux job. I'd like to drop the current job, which is cumbersome and not very relevant, but replace it with a more interesting 32-bit one (e.g., cross-compile to 32-bit ARM) rather than have zero coverage.

**Comment by tupui** [comment]
For the lint job, it would be great to also include the `refguide_check`. This way we can really have a way to skip everything except DOC+CI

**Comment by andyfaff** [comment]
>  a job with pre-releases or the main branch of all our dependencies. Basically that is using "--pre --upgrade --timeout=60 -i $PRE_WHEELS where the env var points at https://pypi.anaconda.org/scipy-wheels-nightly/simple

@rgommers, at the moment this job uses 64 bit BLAS, which I think requires setup.py. Is that correct, or can it use meson now?

**Comment by rgommers** [comment]
> @rgommers, at the moment this job uses 64 bit BLAS, which I think requires setup.py. Is that correct, or can it use meson now?

I'd like to move this to Meson now, and simply add a comment to switch it back over to using ILP64 once that's available with Meson. It's mostly working, just the CLI switch for it and propagating that correctly is not in place. 

We don't seem to have any users for the ILP64 build right now as far as I can tell, and CI for the `setup.py` flavor of this isn't critical; it's unlikely we'll regress on this. We can just add CI back once usage is more streamlined.

**Comment by tupui** [comment]
@andyfaff regarding this job, I already had #15958. Feel free to reopen or make another PR. Just let me know what you intend to do.

**Comment by andyfaff** [comment]
w.r.t the sdist job, how about implementing that in the `linux_meson.test_meson.cp310` entry?

**Comment by andyfaff** [comment]
Sorry, closed by accident due to clumsy fingers.

**Comment by rgommers** [comment]
Oh, I messed up on the four regular Windows jobs: we can't do 32-bit at all, due to lack of compilers. So I think we'll just drop those. So we can reduce it down to 3 jobs, all 64-bit:

- [ ] using `dev.py`, Python 3.9, `'full'` test run, minimum supported `numpy` version
- [ ] using `python setup.py bdist_wheel`, Python 3.9, `'fast'` test run, minimum supported `numpy` version
- [ ] using `python -m build && pip install dist/scipy*.whl`, Python 3.11, `'full'` test run, `-Duse-pythran=false`, (add ILP64 to this later on)

That third one will build via sdist (that's the default behavior for `python -m build`), so that takes care of that too. Does that sound right?

**Comment by tupui** [comment]
@andyfaff do you want to move the Windows part? If not I can try (but might be slow, etc. as I am never touching anything Windows 😉)

**Comment by andyfaff** [comment]
Yes, I'll have a go

**Comment by tupui** [comment]
Awesome thanks Andrew!

**Comment by andyfaff** [comment]
@rgommers what's the correct invocation to turn Pythran off?
`python -m build  -Csetup-args=-Duse_pythran=false` doesn't work.

**Comment by andyfaff** [comment]
Now we just need a 32 bit cross-compile job. @rgommers, can you remember anyone from cross compile issues that might be prepared to make a PR?

**Comment by rgommers** [comment]
Very nice progress, thanks a lot!

> can you remember anyone from cross compile issues that might be prepared to make a PR?

Me I guess. I was already working on it a while back. I have a few days of travel coming up so it may take me a little while though. Let me post what I was thinking:
- Since it's a 32-bit job, we should target `armv7`. That's not available natively on Cirrus, so we may as well do it on GHA (xref our discussion at https://github.com/scipy/scipy/issues/14812#issuecomment-1474837818)
- The above linked comment identifies a couple of GHA actions that can run QEMU (https://github.com/uraimo/run-on-arch-action is another one). 
  - We can't run the whole test suite, but running the fast tests for `scipy.linalg` is possible.
- We have cross-compilation working in https://github.com/conda-forge/scipy-feedstock for `aarch64`, so the invocation works.
- It's easiest to base the cross-compilation environment on conda-forge. The way to set up a non-native environment is given in https://github.com/mesonbuild/meson-python/issues/321#issuecomment-1450231302

After spelling all that out, maybe we should reconsider, and make the cross-compile job that we anyway need target regular `aarch64`, and just port the 32-bit Linux job from Azure to GHA. Reason: QEMU is a pain, and we may still have a coverage gap if we don't run the full test suite.

That Azure job was a pain, but mostly because Azure is a pain and the `&& \` usages make this difficult to read:
```
            set -euo pipefail
            docker pull i386/ubuntu:focal
            docker run -v $(pwd):/scipy i386/ubuntu:focal /usr/bin/linux32 /bin/bash -c "cd scipy && \
            export DEBIAN_FRONTEND=noninteractive && \
            apt-get -y update && \
            apt-get -y install curl python3.9-dev python3.9 python3-distutils pkg-config libpng-dev libjpeg8-dev libfreetype6-dev && \
            curl https://bootstrap.pypa.io/get-pip.py -o get-pip.py && \
            python3.9 get-pip.py && \
            pip3 --version && \
            pip3 install setuptools==59.6.0 wheel numpy==1.21.6 cython==0.29.33 pybind11 pytest pytest-timeout pytest-xdist pytest-env Pillow mpmath pythran pooch && \
            apt-get -y install build-essential gfortran wget && \
            cd .. && \
            mkdir openblas && cd openblas && \
            target=\$(python3.9 ../scipy/tools/openblas_support.py) && \
            cp -r \$target/lib/* /usr/lib && \
            cp \$target/include/* /usr/include && \
            cd ../scipy && \
            python3.9 setup.py install && \
            python3.9 tools/openblas_support.py --check_version $(openblas_version) && \
            python3.9 runtests.py -n --mode=full -- -n auto -s --junitxml=junit/test-results.xml --durations=10 --timeout=60"
```

The job itself is actually quite simple, especially if we remove the cruft here (e.g., don't need the xml stuff nor Pillow or `libpng` et al.). 

One issue is that Ubuntu dropped 32-bit support, so we should use Debian or Alpine.

WDYT?

**Comment by andyfaff** [comment]
It should be relatively easy to translate that job, using containers is straightforward in GHA. If we want a 32-bit job it's preferable to translate that one, rather than have a 32-bit cross-compilation job.

Cross-compilation is a good idea. TBH I was hoping that those who had experienced issues with cross-compilation might be able to share their process of achieving it.
A cross compilation job  would be:

1. Cross compile on a platform that's easy, e.g. build on Ubuntu x86_64 for an Ubuntu/aarch64 host.
2. Fire up QEMU and test the wheel that output from 1.
-or-
3. If we do it on cirrus test the wheel natively in a different task.

**Comment by rgommers** [comment]
I think we want to do both, the cross compilation job is new. The problem with your (2) above is that it's too slow to run the full test suite, and with (3) that there's no 32-bit hardware like `armv7` on Cirrus (I kinda missed that).

**Comment by andyfaff** [comment]
(note for myself, consider https://github.com/marketplace/actions/setup-alpine-linux-environment)

**Comment by rgommers** [comment]
We've achieved the goal here of migrating off of Azure, so I'll close this issue. Thanks @andyfaff, @tupui and everyone else who helped 🎉.

There is still useful content here about adding a new cross compilation CI job, I'll migrate that to its own new issue now.

**Comment by tupui** [comment]
This is awesome! Thanks all 🚀 

Is there some cleaning to do on Azure itself?

**Comment by andyfaff** [comment]
I dont think we need to touch azure itself. There may be an azure webhook that can be disabled in scipy repo settings somewhere.

**Comment by rgommers** [comment]
Before suspending the Azure integration, let's see if @tylerjereddy agrees. We may want to do it now, since it's not hard to reactivate it - but we may also want to wait until after 1.11.0 when we're fairly sure that we don't need the 1.10.x branch anymore.

**Comment by tylerjereddy** [comment]
I don't see a webhook in our settings, looks like it is considered an "App." Think I agree with Ralf on maybe letting the new release coming out before we hard-delete that as well, probably doesn't do much harm for now?

I'm not sure if I'll get another 1.10.x release out, I started having fun with array API stuff, but maybe 1 more that leans on Azure.
